### PR TITLE
Development 3.0: Fix home mount with contain

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

Fix issue #1891 and #1894

### CODE REVIEW

File: src/runtime/engines/singularity/container.go [hunk 1]
```diff
@@ -838,43 +838,45 @@ func (c *container) addBindsMount(system *mount.System) error {
...
 	}
 
-	return nil
+	return source, dest, err
 }
 
-// getHomePaths returns the source and destination path of the requested home mount
-func (c *container) getHomePaths() (source string, dest string, err error) {
-	homeSlice := strings.Split(c.engine.EngineConfig.GetHomeDir(), ":")
+// addHomeStagingDir adds the home staging directory in session directory and
```
Comment (ArangoGutierrez) [inline src/runtime/engines/singularity/container.go]: is this comment incomplete?

**Review by ArangoGutierrez — DISMISSED** [review]
tested

File: src/runtime/engines/singularity/container.go (hunk 1 above)
Comment (cclerget) [reply src/runtime/engines/singularity/container.go]: fixed

**Review by GodloveD — CHANGES_REQUESTED** [review]
This actually doesn't seem to fix #1891 

```
10:01 PM 
ubuntu ~/sandbox
$ singularity --version
singularity version 20758c4d

10:01 PM 
ubuntu ~/sandbox
$ singularity shell --home /data my_container
Singularity :/home/david/sandbox> pwd
/home/david/sandbox
```
In Singularity 2.6 that `pwd` command returns `/data`

**Comment by cclerget** [comment]
@GodloveD Actually the code mount custom home but doesn't chdir into it since the current working directory is mounted with :
`cd $HOME/test && singularity shell --home /tmp/home my_container`

For the behaviour you described :

`cd $HOME/test && singularity shell --home /tmp/home:$HOME my_container`

I'm wondering if this is a bug to mount cwd with `--home` as long as cwd doesn't conflict with container home ?

**Comment by GodloveD** [comment]
Oh I see.  Maybe this is just a bug in 2.6 then and we should _not_ `cd` to `$HOME` when we startup.  After all, we don't `cd` to `$HOME` when we start singularity without the `--home` option.  So that makes sense.  

But now I think the behavior is confusing the other way around.  Why do we `cd` to `$HOME` with this option argument pair?

```
--home /tmp/home:$HOME
```

Shouldn't we just stay in the `cwd` in that case too?  To be clear, the difference between these two commands doesn't make sense to me.

```
07:48 AM 
ubuntu ~/go/src/github.com/singularityware/singularity
$ cd ~/test

07:48 AM 
ubuntu ~/test
$ singularity exec --home /data:/mnt ~/sandbox/my_container pwd
/home/david/test

07:48 AM 
ubuntu ~/test
$ singularity exec --home /data:$HOME ~/sandbox/my_container pwd
/home/david
```

**Comment by cclerget** [comment]
It depends where you are when you run Singularity, in your example you have `CWD=/home/david/test`, if you run with `--home /data:/home/david`, Singularity couldn't chdir to cwd because `/data` doesn't contain `test` directory path in order to bind it, that's why Singularity chdir goes into `/home/david`, if you create `/data/test` it should work with `CWD=/home/david/test`.

I agree this is confusing, and that may require to define a default behaviour after discussion

**Comment by jmstover** [comment]
I thought current behavior was this:

1) `cd` to `$CWD` on host if possible  (i.e. we try mounting `$CWD`)
2) `cd` to `$HOME`
3) `cd` to `/`

So, if the host `$CWD` location is available, it wins; Then we fall back to `$HOME`; and finally `/`

**Comment by GodloveD** [comment]
@jmstover I think that _should_ be the behavior.  But it doesn't look like it is the behavior?

**Comment by jmstover** [comment]
This is what I get with 2.6.0

```
$ cd /tmp
$ /usr/local/singularity/2.6.0/bin/singularity exec ~/tmp/sing/centos7.img pwd
/tmp
$ /usr/local/singularity/2.6.0/bin/singularity exec -c ~/tmp/sing/centos7.img pwd
/home/jason
$ /usr/local/singularity/2.6.0/bin/singularity exec -c --no-home ~/tmp/sing/centos7.img pwd
WARNING: Could not chdir to home: /home/jason
/
```

So, $CWD, $HOME, then /

**Comment by bauerm97** [comment]
Just want to chime in and say that I agree with @jmstover here. I think that behavior should be:

1. `cd` to `$CWD`
2. `cd` to `$HOME`
3. `cd` to `/`

**Comment by jmstover** [comment]
@GodloveD  ... This looks like it's working correctly to me:

```
$ cd /bin2
$ /usr/local/singularity/3.0.0-dev/bin/singularity exec ~/tmp/sing/alpine.sif pwd
/bin2
$ cd /tmp/
$ /usr/local/singularity/3.0.0-dev/bin/singularity exec -c ~/tmp/sing/alpine.sif pwd
/home/jason
$ /usr/local/singularity/3.0.0-dev/bin/singularity exec -c --no-home ~/tmp/sing/alpine.sif pwd
/
```

**Comment by GodloveD** [comment]
I think you are right @jmstover.  In my second test above:

```
07:48 AM 
ubuntu ~/test
$ singularity exec --home /data:$HOME ~/sandbox/my_container pwd
/home/david
```

the `~/test` directory is no longer available and so it defaults to `$HOME`.  Thanks for clearing that up.

**Review by GodloveD — DISMISSED** [review]

**Comment by GodloveD** [comment]
@cclerget you've got two 👍 reviews now.  If you would like to resolve the conflicts and then merge, please feel free.  😸

**Review by jmstover — APPROVED** [review]

**Review by GodloveD — APPROVED** [review]
lgtm.  Thanks @ArangoGutierrez!

EDIT:  Whoops!  Put this approval and comment on the wrong PR I think.  This one is already good to go.

**Pull request merged**